hathor/util.py

```diff
@@ -503,3 +503,157 @@ def progress(iter_tx: Iterator['BaseTransaction'], *, log: Optional['structlog.s
+            except StopIteration:
+                self._it = None
+                return None
+        return self._head
+
+    def __iter__(self) -> Iterator[T]:
+        return self
+
+    def __next__(self) -> T:
+        if self._head is not None:
+            (x,), self._head = self._head, None
+            return x
+        elif self._it is not None:
+            try:
+                return next(self._it)
+            except StopIteration:
+                self.it = None
+                raise
+        else:
+            raise StopIteration()
+
+    def peek(self) -> T:
+        x = self._peek()
+        if x is None:
+            raise ValueError('iterator was exhausted')
+        y, = x
+        return y
+
+    def __bool__(self) -> bool:
+        return self._peek() is not None
+
+
+def _identity(x: T) -> T:
+    return x
+
+
+class sorted_merger(Iterator[T]):
+    """ Adaptor class to merge multiple sorted iterators into a single iterator that is also sorted.
+
+    Note: for this adaptor to work as expected the input iterators have to already be sorted, but if they aren't, the
+    resulting iterator won't crash, or unexpecetdly stop working, however it will not be sorted.
+
+    A custom key function can be supplied to customize the sort order, and the
+    reverse flag can be set to request the result in descending order.
+
+    The implemented logic is really simple:
+
+    - Peek the next element in each iterator, and yield from the "smallest" one (according to the key function and the
+      reversed flag), when an iterator is exhausted it's just removed from the list until the list is empty, in which
+      point the resulting iterator will stop.
+
+    For example:
+
+    >>> list(sorted_merger([1,3,4,100,101,105], [104], [2,50,99,106]))
+    [1, 2, 3, 4, 50, 99, 100, 101, 104, 105, 106]
+
+    For descending order, use reverse=True
+
+    >>> list(sorted_merger([105,101,100,4,3,1], [104], [106,99,50,2], reverse=True))
+    [106, 105, 104, 101, 100, 99, 50, 4, 3, 2, 1]
+
+    But using a negating key also works
+
+    >>> list(sorted_merger([105,101,100,4,3,1], [104], [106,99,50,2], key=lambda i: -i))
+    [106, 105, 104, 101, 100, 99, 50, 4, 3, 2, 1]
+
+    Empty stuff will just yield empty stuff
+
+    >>> list(sorted_merger([]))
+    []
+
+    >>> list(sorted_merger())
+    []
+
+    All elements will eventually be yielded, it doesn't matter if they are "repeated"
+
+    >>> list(sorted_merger([], [1,1,1], [1,1], [1,1,1,1]))
+    [1, 1, 1, 1, 1, 1, 1, 1, 1]
+
+    Even if they are not sorted, they will still be yielded eventually, but there are no guarantees about the order
+    they will come out
+
+    >>> list(sorted_merger([1,2,3],[4,3,2]))
+    [1, 2, 3, 4, 3, 2]
+    """
+
+    def __init__(self, *iterators: Iterator[T], key: Optional[Callable[[T], Any]] = None,
+                 reverse: bool = False) -> None:
+        self._iterators = [peekable(it) for it in iterators]
```

Comment:
While I agree with that, I think it's much easier to use a `peekable` than to juggle temporary values, specially because I'd have to juggle them in a very similar way for more than one case. I considered adding a dependency that would provide this `peekable` (more-itertools provides _exactly _ that), but I think it's simple enough that it could live in `hathor.util`.